### PR TITLE
wherein we suggest float for integer literals where a float was expected

### LINKED ISSUE #53280: "integral variable" should suggest how to make literal a float
The following code:

```rust
fn foo(x: f64) {}

fn main() {
    foo(10);
}
```

Produces the following error message:

```rust
error[E0308]: mismatched types
 --> src/main.rs:4:9
  |
4 |     foo(10);
  |         ^^ expected f64, found integral variable
  |
  = note: expected type `f64`
             found type `{integer}`
```

This error message tells you that the types are mismatched, but doesn't fully explain why. To fix it, you need to use a floating point literal instead of an integer literal. This isn't clear from the error message.

If you're new to programming or if you come from a language that doesn't draw the same distinction between integer and floating point literals, it may not be clear why the literal `10` is not the same as `10.0`. 

It would be great if we could add something in our error message that would suggest one or more of the valid ways to fix this:

- `10.0`
- `10.`
- `10f64` - type could be `f32` in another context

### PR DESCRIPTION
**Pull request by zackmdavis**
@sunjay pointed out that this is a nice thing that we could do.

Resolves #53280.

r? @estebank

### CODE REVIEW

**Comment by sunjay** [comment]
This is awesome! Thank you for doing this so fast! :smile: 

Will this code work if the integer literal has underscores in it? What about negative numbers? The full spec of integer literals is here: https://doc.rust-lang.org/reference/tokens.html#integer-literals

My personal opinion is that you don't need to support all of them, but we should at least support literals in the format `1_000_000` since that is equivalent to `1000000`. Negative integers are also a must.

Thanks again! This seems like a small thing, but I think it's a big quality of life improvement for people who are new to Rust! :smiley:

**Comment by GuillaumeGomez** [comment]
@zackmdavis Please add tests for all cases that @sunjay gave. 😆

File: src/librustc/ty/error.rs [hunk 1]
```diff
@@ -250,6 +250,21 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
                     db.note("no two closures, even if identical, have the same type");
                     db.help("consider boxing your closure and/or using it as a trait object");
                 }
+                match (&values.found.sty, &values.expected.sty) { // Issue #53280
```
Comment (estebank) [inline src/librustc/ty/error.rs]: It's nice to provide a bit of context in the comments without the need to open a browser :)

As a rule, I add a one sentence long description of what the code is doing (or rather, why), with the issue number at the end for _extra_ context.

**Review by estebank — APPROVED** [review]
Love the change.

r=me once travis is happy.

File: src/librustc/ty/error.rs (hunk 1 above)
Comment (zackmdavis) [reply src/librustc/ty/error.rs]: I agree that issue numbers are merely an extra reference for the curious software-archeologist and not a substitute for explanatory comments where explanatory comments are merited, but I feel like "use a float literal" is pretty self-explanatory? I do write explanatory comments where I think it necessary (recent examples: [1](https://github.com/rust-lang/rust/blame/0aa8d0320266b5579428312095fe49af05ada972/src/librustc/hir/lowering.rs#L1772-L1777) [2](https://github.com/rust-lang/rust/blame/0aa8d0320266b5579428312095fe49af05ada972/src/libsyntax/parse/parser.rs#L6047-L6049)).

**Pull request merged**